cmd/terraform-provider-qbec/main.go

```diff
@@ -0,0 +1,294 @@
+							fldExtCodeVars: {
+								Type:        schema.TypeMap,
+								Description: "external variables to set as code variables",
+								Elem:        &schema.Schema{Type: schema.TypeString},
+								Optional:    true,
+							},
+							fldTLAVars: {
+								Type:        schema.TypeMap,
+								Description: "TLA variables to set as strings",
+								Elem:        &schema.Schema{Type: schema.TypeString},
+								Optional:    true,
+							},
+							fldTLACodeVars: {
+								Type:        schema.TypeMap,
+								Description: "TLA variables to set as code variables",
+								Elem:        &schema.Schema{Type: schema.TypeString},
+								Optional:    true,
+							},
+						},
+						ReadContext: evalJsonnet,
+					},
+				},
+			}
+		},
+	})
+}
+
+// since we need to change the workdir, only one eval can happen at any one time :(
+var globalLock sync.Mutex
+
+func evalJsonnet(ctx context.Context, data *schema.ResourceData, _ interface{}) (res diag.Diagnostics) {
+	globalLock.Lock()
+	defer globalLock.Unlock()
+
+	{
+		root := data.Get(fldRoot).(string)
+		wd, err := os.Getwd()
+		if err != nil {
+			return diag.FromErr(errors.Wrap(err, "get working directory"))
+		}
+		if err := os.Chdir(root); err != nil {
+			return diag.FromErr(errors.Wrapf(err, "change working directory to %s", root))
+		}
+		defer func() { _ = os.Chdir(wd) }()
+	}
+
+	var file, code string
+
+	{
+		file0, fileFound := data.GetOk("file")
+		code0, codeFound := data.GetOk("code")
+
+		switch {
+		case !fileFound && !codeFound:
+			return diag.FromErr(fmt.Errorf("one of '%s' or '%s' attributes must be set", fldCode, fldFile))
+		case fileFound && codeFound:
+			return diag.FromErr(fmt.Errorf("cannot set both '%s' and '%s'", fldCode, fldFile))
+		case fileFound:
+			file = file0.(string)
+		default:
+			code = code0.(string)
+			if strings.Trim(code, " \t\n\r") == "" {
+				return diag.FromErr(fmt.Errorf("%s string is empty", fldCode))
+			}
+		}
+	}
+
+	var format string
+	{
+		format0, ok := data.GetOk(fldFormat)
+		if !ok {
+			format = fmtJSON
+		} else {
+			format = format0.(string)
+		}
+		if format == "" {
+			format = fmtJSON
+		}
+		switch format {
+		case fmtJSON, fmtYAML, fmtMultiYAML:
+		default:
+			return diag.FromErr(fmt.Errorf("invalid %s field, want be one of %s, %s, or %s, got %s", fldFormat, fmtJSON, fmtYAML, fmtMultiYAML, format))
+		}
+	}
+
+	libPaths, err := getLibPaths(data)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	vars, err := getVariables(data)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	bc := eval.BaseContext{
+		LibPaths: libPaths,
+		Vars:     vars,
+	}
+
+	var jsonString string
+
+	if file != "" {
+		jsonString, err = eval.File(file, bc)
+	} else {
+		jsonString, err = eval.Code("<inline-code>", vm.MakeCode(code), bc)
+	}
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	switch format {
+	case fmtJSON:
+		if err := data.Set(fldComputedResult, jsonString); err != nil {
+			return diag.FromErr(err)
+		}
+	default:
```

Comment:
well, this is the json string as returned by a `jsonnet eval` and doesn't have to be an object - it could be simply a json-quoted string. Maybe we could support a `raw` output format that asserts that the output of the eval is of type string and unescapes the string to return its raw form instead?

Reply:
actually the jsonnet code could just return a string, the `result` attribute will be the json-quoted version of the string and the caller should be able to use the  `jsondecode` function in terraform to get the raw value of the string

Reply:
i should add an example to make this use-case clear (e.g. use `std.manifestIni` to return the contents of an INI file)